File: Ejercicio1/1.py
from datetime import datetime
from typing import List

# Clase Autor
class Autor:
    def __init__(self, nombre: str, nacionalidad: str):
        self.nombre = nombre
        self.nacionalidad = nacionalidad
        self.librosEscritos = []

# Clase Libro
class Libro:
    def __init__(self, isbn: str, titulo: str, anoPublicacion: int, cantidadEjemplares: int, autor: Autor):
        self.isbn = isbn
        self.titulo = titulo
        self.anoPublicacion = anoPublicacion
        self.cantidadEjemplares = cantidadEjemplares
        self.autor = autor

# Clase Usuario
class Usuario:
    def __init__(self, nombre: str, direccion: str, telefono: str):
        self.nombre = nombre
        self.direccion = direccion
        self.telefono = telefono
        self.librosPrestados = []

# Clase Prestamo
class Prestamo:
    def __init__(self, libro: Libro, usuario: Usuario, fechaPrestamo: datetime):
        self.libro = libro
        self.usuario = usuario
        self.fechaPrestamo = fechaPrestamo
        self.fechaDevolucion = None

    def devolver_Libro(self, fechaDevolucion: datetime):
        self.fechaDevolucion = fechaDevolucion
        # Eliminar el libro de la lista de libros prestados del usuario
        if self.libro in self.usuario.librosPrestados:
            self.usuario.librosPrestados.remove(self.libro)

# Clase Biblioteca
class Biblioteca:
    def __init__(self):
        self.libros = []
        self.autores = []
        self.usuarios = []
        self.prestamos = []

    def RegistrarAutor(self, autor: Autor):
        self.autores.append(autor)

    def RegistrarLibro(self, libro: Libro):
        self.libros.append(libro)

    def RegistrarUsuario(self, usuario: Usuario):
        self.usuarios.append(usuario)

    def PrestarLibro(self, isbn: str, nombreUsuario: str, fechaPrestamo: datetime):
        libro = None
        usuario = None

        # Buscar libro por ISBN
        for l in self.libros:
            if l.isbn == isbn:
                libro = l
                break

        # Buscar usuario por nombre
        for u in self.usuarios:
            if u.nombre == nombreUsuario:
                usuario = u
                break

        if libro and usuario:
            prestamo = Prestamo(libro, usuario, fechaPrestamo)
            self.prestamos.append(prestamo)
            usuario.librosPrestados.append(libro)
            print("Prestamo exitoso ;D")
        else:
            print("Libro o usuario no encontrado =(")

    def DevolverLibro(self, isbn: str, nombreUsuario: str, fechaDevolucion: datetime):
        prestamo = None
        # Buscar prestamo por ISBN y nombre de usuario
        for p in self.prestamos:
            if p.libro.isbn == isbn and p.usuario.nombre == nombreUsuario and p.fechaDevolucion is None:
                prestamo = p
                break

        if prestamo:
            prestamo.devolver_Libro(fechaDevolucion)
            print("Devolucion exitosa ;D")
        else:
            print("Prestamo no encontrado o ya fue devuelto")


# Codigo principal

biblioteca = Biblioteca()
    
# Registrar autor
autor1 = Autor("La llamada de Cthulhu", "Estados Unidos")
biblioteca.RegistrarAutor(autor1)
    
# Registrar libro
libro1 = Libro("123456789", "El color que cayo del cielo", 1967, 5, autor1)
biblioteca.RegistrarLibro(libro1)
autor1.librosEscritos.append(libro1.titulo)
    
# Registrar usuario
usuario1 = Usuario("Pedro Perez", "Av. Principal 553", "555-1234")
biblioteca.RegistrarUsuario(usuario1)
    
# Prestar libro
fechaPrestamo = datetime.strptime("2024-04-12", "%Y-%m-%d")
biblioteca.PrestarLibro("123456789", "Pedro Perez", fechaPrestamo)
    
# Devolver libro
fechaDevolucion = datetime.strptime("2024-04-15", "%Y-%m-%d")
biblioteca.DevolverLibro("123456789", "Pedro Perez", fechaDevolucion)

File: Ejercicio1/1.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

// Clase Autor
class Autor 
{
public:
    string nombre;
    string nacionalidad;
    vector<string> librosEscritos;

    Autor(string nombre, string nacionalidad) 
    {
        this->nombre = nombre;
        this->nacionalidad = nacionalidad;
    }
};


class Libro 
{

public:
    string isbn;
    string titulo;
    int anioPublicacion;
    int cantidadEjemplares;
    Autor* autor;

    Libro(string isbn, string titulo, int anioPublicacion, int cantidadEjemplares, Autor* autor) {
        this->isbn = isbn;
        this->titulo = titulo;
        this->anioPublicacion = anioPublicacion;
        this->cantidadEjemplares = cantidadEjemplares;
        this->autor = autor;
    }
};


class Usuario 
{
public:
    string nombre;
    string direccion;
    string telefono;
    vector<Libro*> librosPrestados;

    Usuario(string nombre, string direccion, string telefono) 
    {
        this->nombre = nombre;
        this->direccion = direccion;
        this->telefono = telefono;
    }
};


class Prestamo 
{
public:
    Libro* libro;
    Usuario* usuario;
    string fechaPrestamo;
    string fechaDevolucion;

    Prestamo(Libro* libro, Usuario* usuario, string fechaPrestamo) 
    {
        this->libro = libro;
        this->usuario = usuario;
        this->fechaPrestamo = fechaPrestamo;
        this->fechaDevolucion = "";
    }

    void devolverLibro(string fechaDevolucion) 
    {
        this->fechaDevolucion = fechaDevolucion;
        // Eliminar el libro de la lista de libros prestados del usuario
        for (auto it = usuario->librosPrestados.begin(); it != usuario->librosPrestados.end(); ++it) 
        {
            if (*it == libro) 
            {
                usuario->librosPrestados.erase(it);
                break;
            }
        }
    }
};


class Biblioteca 
{
public:
    vector<Libro*> libros;
    vector<Autor*> autores;
    vector<Usuario*> usuarios;
    vector<Prestamo*> prestamos;

    void registrarAutor(Autor* autor) 
    {
        autores.push_back(autor);  //PUSHBACK PARA METER ALGO A UN VECTOR
    }

    void registrarLibro(Libro* libro)  
    {
        libros.push_back(libro); //PUSHBACK PARA METER ALGO A UN VECTOR
    }

    void registrarUsuario(Usuario* usuario) 
    {
        usuarios.push_back(usuario); //PUSHBACK PARA METER ALGO A UN VECTOR
    }

    void prestarLibro(string isbn, string nombreUsuario, string fechaPrestamo) 
    {
        Libro* libro = nullptr;
        Usuario* usuario = nullptr;

                                     
        for (auto l : libros)    //SE BUSCA EL LIBRO POS ISBN
        {
            if (l->isbn == isbn) 
            {
                libro = l;
                break;
            }
        }


        for (auto u : usuarios) 
        {
            if (u->nombre == nombreUsuario) 
            {
                usuario = u;
                break;
            }
        }

        if (libro && usuario) 
        {
            Prestamo* prestamo = new Prestamo(libro, usuario, fechaPrestamo);
            prestamos.push_back(prestamo);
            usuario->librosPrestados.push_back(libro);
            cout << "Prestamo exitosamente realizado ;D" << endl;
        } 
        else 
        {
            cout << "Libro o usuario no encontrado =(" << endl;
        }
    }

    void devolverLibro(string isbn, string nombreUsuario, string fechaDevolucion) 
    {
        Prestamo* prestamo = nullptr;
        // BUSCAMOS EL PRESTAMO POR ISBN Y NOMBRE DE USUARIO
        for (auto p : prestamos) 
        {
            if (p->libro->isbn == isbn && p->usuario->nombre == nombreUsuario && p->fechaDevolucion.empty()) 
            { 
                prestamo = p;
                break;
            }
        }

        if (prestamo) 
        {
            prestamo->devolverLibro(fechaDevolucion);
            cout << "Devolucion exitosamente realizada ;D" << endl;
        } 
        else 
        {
            cout << "Prestamo no encontrado o ya fue devuelto" << endl;
        }
    }
};

int main() 
{

    Biblioteca biblioteca;
    Autor* autor1 = new Autor("La llamada de Cthulhu", "Estados Unidos");
    biblioteca.registrarAutor(autor1);


    Libro* libro1 = new Libro("123456789", "El color que cayo del cielo", 1967, 5, autor1);
    biblioteca.registrarLibro(libro1);
    autor1->librosEscritos.push_back(libro1->titulo);

    Usuario* usuario1 = new Usuario("Pedro Perez", "Av. Principal 553", "555-1234");
    biblioteca.registrarUsuario(usuario1);


    biblioteca.prestarLibro("123456789", "Pedro Perez", "2024-04-12");
    biblioteca.devolverLibro("123456789", "Pedro Perez", "2024-04-15");

    return 0;
}
